Autogenerate syncing of `IsAutofocusProcessed` and `UserDidInteractWithPage` data

beidson opened the pull request:
#### f8feef1a83396879cc2755f94cb25c2871fd0a27
<pre>
Autogenerate syncing of `IsAutofocusProcessed` and `UserDidInteractWithPage` data
<a href="https://rdar.apple.com/141378639">rdar://141378639</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=284573">https://bugs.webkit.org/show_bug.cgi?id=284573</a>

Reviewed by Alex Christensen.

Adding these two types at the same time exposed the next issue that needed to be solved
in the generator script - having multiple copies of the same type in the variant.

We handle this by:
1 - Switching from direct initialization of the variant to using emplace() with an index
2 - Switching from std::get&lt;type&gt; to std::get&lt;(integer index&gt;)
3 - Giving explicit integer values to the `enum class` member that aligns with the type&apos;s index

* Source/WebCore/Scripts/generate-process-sync-data.py:
(SyncedData.__init__):
(generate_process_sync_client_header):
(generate_process_sync_client_impl):
(generate_process_sync_data_header):
(generate_document_synched_data_header):
(generate_document_synched_data_impl):
(generate_process_sync_data_serialiation_in):
(sort_datas_for_variant_order):
(main):
(sorted_qualified_types): Deleted.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::noteUserInteractionWithMediaElement):
(WebCore::Document::flushAutofocusCandidates):
* Source/WebCore/dom/Document.h:
(WebCore::Document::isAutofocusProcessed const): Deleted.
(WebCore::Document::setAutofocusProcessed): Deleted.
(WebCore::Document::setUserDidInteractWithPage): Deleted.
(WebCore::Document::userDidInteractWithPage const): Deleted.
* Source/WebCore/dom/Element.cpp:
(WebCore::shouldAutofocus):
* Source/WebCore/dom/UserGestureIndicator.cpp:
* Source/WebCore/html/HTMLDialogElement.cpp:
(WebCore::HTMLDialogElement::runFocusingSteps):
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::runPopoverFocusingSteps):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::shouldAskForNavigationConfirmation):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::keyEvent):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setUserDidInteractWithPage):
(WebCore::Page::userDidInteractWithPage const):
(WebCore::Page::setAutofocusProcessed):
(WebCore::Page::autofocusProcessed const):
(WebCore::Page::updateProcessSyncData):
* Source/WebCore/page/Page.h:
* Source/WebCore/page/ProcessSyncData.in:

Canonical link: <a href="https://commits.webkit.org/287820@main">https://commits.webkit.org/287820@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/07e10161bd372364c9bd7775ab4330ef70901b72

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80943 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/470 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34887 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85474 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31931 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83054 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/483 "Build is being retried. Recent messages:OS: Sequoia (15.2), Xcode: 16.2; Skipping applying patch since patch_id isn't provided; Checked out pull request; compiling") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8271 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/63212 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/20979 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84012 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/130/builds/483 "Build is being retried. Recent messages:OS: Sequoia (15.2), Xcode: 16.2; Skipping applying patch since patch_id isn't provided; Checked out pull request; compiling") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73678 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43510 "Passed tests") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/130/builds/483 "Build is being retried. Recent messages:OS: Sequoia (15.2), Xcode: 16.2; Skipping applying patch since patch_id isn't provided; Checked out pull request; compiling") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27835 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30389 "Built successfully") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/130/builds/483 "Build is being retried. Recent messages:OS: Sequoia (15.2), Xcode: 16.2; Skipping applying patch since patch_id isn't provided; Checked out pull request; compiling") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28384 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86909 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8175 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/5782 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71511 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8352 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69510 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70745 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17613 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14776 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13706 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8136 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/13659 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7973 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11493 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9781 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/Scripts/generate-process-sync-data.py:
```diff
@@ -186,12 +187,14 @@ def generate_process_sync_client_impl(synched_datas):
         result.append('void ProcessSyncClient::broadcast%sToOtherProcesses(const %s& data)' % (data.name, data.fully_qualified_type))
         result.append('{')
-        result.append('    broadcastProcessSyncDataToOtherProcesses({ ProcessSyncDataType::%s, { data }});' % data.name)
+        result.append('    ProcessSyncDataVariant dataVariant;')
```
Would still be great to have a test so we can see what this does to output and acceptable input.

beidson replied:
I have the new files in a directory on my machine ready to go after this commit is ready to merge.

achristensen07 commented on Source/WebCore/Scripts/generate-process-sync-data.py:
```diff
@@ -348,15 +328,15 @@ def generate_document_synched_data_impl(synched_datas):
         lowercase_name = data.name[0].lower() + data.name[1:]
         result.append('    case ProcessSyncDataType::%s:' % data.name)
-        result.append('        %s = std::get<%s>(data.value);' % (lowercase_name, data.fully_qualified_type))
+        result.append('        %s = std::get<(int)ProcessSyncDataType::%s>(data.value);' % (lowercase_name, data.name))
```
This isn't great.  Can we use WTF::enumToUnderlyingType instead?

achristensen07 commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -5346,8 +5346,10 @@ void Document::noteUserInteractionWithMediaElement()
 
-    if (!protectedTopDocument()->userDidInteractWithPage())
-        return;
+    if (RefPtr page = protectedPage()) {
```
I think if there's no page we should also early return.

achristensen07 commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -5657,15 +5659,16 @@ void Document::clearAutofocusCandidates()
 {
     ASSERT(isTopDocument());
-    if (m_isAutofocusProcessed)
+    RefPtr page = protectedPage();
```
I think this is supposed to call this->page() instead of protectedPage for some pedantic reason I don't understand.  Not sure though.

beidson replied:
Pretty sure they're functionally equivalent:

`inline Page* Document::page() const
{
    return m_frame ? m_frame->page() : nullptr;
}

inline RefPtr<Page> Document::protectedPage() const
{
    return page();
}`

achristensen07 commented on Source/WebCore/page/Page.cpp:
```diff
@@ -817,11 +817,39 @@ DOMAudioSessionType Page::audioSessionType() const
 
+void Page::setUserDidInteractWithPage(bool didInteract)
+{
+    m_documentSyncData->userDidInteractWithPage = didInteract;
```
Early return if this doesn't change the value?  Same in setAudioSessionType

achristensen07 commented on Source/WebCore/page/Page.cpp:
```diff
@@ -817,11 +817,39 @@ DOMAudioSessionType Page::audioSessionType() const
     switch (data.type) {
     case ProcessSyncDataType::MainFrameURLChange:
-        setMainFrameURL(std::get<URL>(data.value));
+        setMainFrameURL(std::get<(int)ProcessSyncDataType::MainFrameURLChange>(data.value));
```
WTF::enumToUnderlyingType instead of c-style cast.

achristensen07 commented on Source/WebCore/page/ProcessSyncData.in:
```diff
@@ -58,3 +58,5 @@
 
 MainFrameURLChange : URL [Header=<wtf/URL.h>]
 AudioSessionType : WebCore::DOMAudioSessionType [DocumentSyncData Conditional=ENABLE(DOM_AUDIO_SESSION) Header="DOMAudioSession.h"]
+IsAutofocusProcessed : bool [DocumentSyncData]
```
I think it would be a little nicer to have the attributes before the other stuff, but no strong opinion.

beidson committed:
Autogenerate syncing of `IsAutofocusProcessed` and `UserDidInteractWithPage` data
rdar://141378639
https://bugs.webkit.org/show_bug.cgi?id=284573

Reviewed by Alex Christensen.

Adding these two types at the same time exposed the next issue that needed to be solved
in the generator script - having multiple copies of the same type in the variant.

We handle this by:
1 - Switching from direct initialization of the variant to using emplace() with an index
2 - Switching from std::get<type> to std::get<(integer index>)
3 - Giving explicit integer values to the `enum class` member that aligns with the type's index

* Source/WebCore/Scripts/generate-process-sync-data.py:
(SyncedData.__init__):
(generate_process_sync_client_header):
(generate_process_sync_client_impl):
(generate_process_sync_data_header):
(generate_document_synched_data_header):
(generate_document_synched_data_impl):
(generate_process_sync_data_serialiation_in):
(sort_datas_for_variant_order):
(main):
(sorted_qualified_types): Deleted.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::noteUserInteractionWithMediaElement):
(WebCore::Document::flushAutofocusCandidates):
* Source/WebCore/dom/Document.h:
(WebCore::Document::isAutofocusProcessed const): Deleted.
(WebCore::Document::setAutofocusProcessed): Deleted.
(WebCore::Document::setUserDidInteractWithPage): Deleted.
(WebCore::Document::userDidInteractWithPage const): Deleted.
* Source/WebCore/dom/Element.cpp:
(WebCore::shouldAutofocus):
* Source/WebCore/dom/UserGestureIndicator.cpp:
* Source/WebCore/html/HTMLDialogElement.cpp:
(WebCore::HTMLDialogElement::runFocusingSteps):
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::runPopoverFocusingSteps):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::shouldAskForNavigationConfirmation):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::keyEvent):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setUserDidInteractWithPage):
(WebCore::Page::userDidInteractWithPage const):
(WebCore::Page::setAutofocusProcessed):
(WebCore::Page::autofocusProcessed const):
(WebCore::Page::updateProcessSyncData):
* Source/WebCore/page/Page.h:
* Source/WebCore/page/ProcessSyncData.in:

Canonical link: https://commits.webkit.org/287820@main

rkirsling commented on Source/WebCore/Scripts/generate-process-sync-data.py:
```diff
@@ -186,12 +187,14 @@ def generate_process_sync_client_impl(synched_datas):
         result.append('{')
-        result.append('    broadcastProcessSyncDataToOtherProcesses({ ProcessSyncDataType::%s, { data }});' % data.name)
+        result.append('    ProcessSyncDataVariant dataVariant;')
+        result.append('    dataVariant.emplace<enumToUnderlyingType(ProcessSyncDataType::%s)>(data);' % data.name)
```
Looks like `enumToUnderlyingType` requires inclusion of EnumTraits.h. https://github.com/WebKit/WebKit/pull/37956
